google/cloud/bigtable/client.py

```diff
@@ -510,20 +520,49 @@ async def read_rows_sharded(
         self,
         query_list: list[ReadRowsQuery] | list[dict[str, Any]],
         *,
-        limit: int | None,
-        operation_timeout: int | float | None = 60,
+        operation_timeout: int | float | None = None,
         per_request_timeout: int | float | None = None,
-    ) -> ReadRowsIterator:
+    ) -> list[Row]:
         """
-        Runs a sharded query in parallel
+        Runs a sharded query in parallel, then return the results in a single list.
+        Results will be returned in the order of the input queries.
+
+        This function is intended to be run on the results on a query.shard() call:
 
-        Each query in query list will be run concurrently, with results yielded as they are ready
-        yielded results may be out of order
+        ```
+        table_shard_keys = await table.sample_keys()
+        query = ReadRowsQuery(...)
+        shard_queries = query.shard(table_shard_keys)
+        results = await table.read_rows_sharded(shard_queries)
+        ```
 
         Args:
             - query_list: a list of queries to run in parallel
-        """
-        raise NotImplementedError
+        Raises:
+            - ShardedReadRowsExceptionGroup: if any of the queries failed
+            - ValueError: if the query_list is empty
+        """
```

Comment:
Do we want to raise an error if any of the shard queries overlap? Or is it ok to get duplicate rows?

Reply:
Yeah assuming they use the query.shard() function, that should be the case. But this method allows passing in a generic list of queries, so users may pass in overlapping queries, right?

Reply:
You are right that its possible. I think we should avoid this situation, but not by throwing an error. I think we should make it impossible to happen. Perhaps we can do the following:

Create a Batch fetching context that end users create. The context will automatically call SampleRowKeys and cache the result. And maybe refresh it every X minutes.

The end user then interact with this object by passing it lists of keys and ranges that the context shards

Reply:
And then move the `read_rows_sharded(unsharded_query)` function onto the context object? Or something else? I'd be a bit hesitant to add more background tasks if we can avoid it, but we can probably work something out.

Another option that would be very simple to add would be to make `query.shard` return a custom `ShardedQuery` object that just wraps the query list, and then only accept that as input for `read_rows_sharded`. Or even simpler, just make define it as a type alias

Is this something we can create an issue for and address after the first alpha, or do you want it resolved before merging this?

Reply:
Ok, I made a custom type for `ShardedQueries`, which should discourage people from passing their own custom queries. We can discuss more advanced changes later. Let me know what you think